### DISCUSSION TITLE
where is waydroid kernel located?

### DISCUSSION THREAD

**Discussion by realyukii** [waydroid/waydroid#1552] [Q&A]
When exploring the mounted filesystem at `/var/lib/waydroid/rootfs`, I tried to locate the Linux kernel by searching for a vmlinuz file, but I couldn't find one. I'm curious about how the kernel is booted and utilized within the Waydroid system. I looked through the documentation but didn't find any relevant information.

It would be great if the documentation explained how Waydroid works under the hood or provided references to help build a better understanding of this area.

**Comment by realyukii** [comment]
I heard that Waydroid is container-based. Is it similar to Docker? But Docker doesn’t run its own kernel; it uses the host's kernel instead, right? Does that mean Waydroid operates in a similar way to Docker?

**Reply by realyukii** [reply]
But `uname -a` on adb shell doesn't seem like a kernel host: `Linux waydroid 6.10.6-arch1-1 #1 SMP PREEMPT_DYNAMIC Mon, 19 Aug 2024 17:02:39 +0000 x86_64`

**Reply by realyukii** [reply]
I discovered that Waydroid runs on top of LXC, so it seems there might not be a separate Android kernel in Waydroid. Instead, it might be modifying the host kernel in some way, though I'm not sure how this is done.

I tried listing the LXC containers with `lxc-ls -f` as root, but the command didn’t return anything. It’s strange.

[`lsnet(1)`](https://wizardzines.com/comics/namespaces/)

**Comment by jimed-rand** [comment]
Anyway, Waydroid uses `binder` and `ashmem` modules which required for Waydroid to communicate with the host's kernel. As you can see, If you'd check it via CPU-Z or AIDA64, the kernel version will describe the host kernel we use on our Linux sessions.

And severe distro like I'd use now (openSUSE Tumbleweed) needs `anbox-modules` which loaded with `binder` and `ashmem` module and for stability, I'd use an `kernel-longterm` to prevent comms problem with the module Waydroid needs. And for installation guide, I'd follow the #1463 for install it and I help maintain the @runa-chin's package in OBS (openSUSE Build Service).

**Comment by aleasto** [comment]
Nowhere. Waydroid is not a virtual machine; we run on top of your host kernel.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/waydroid/waydroid/discussions/1552#discussioncomment-12543517

@aleasto, does the AOSP kernel being compiled with Bionic, not GLibC, not prevent this (or is my premise nonsensical, unbenknownst to me)?

**Reply by aleasto** [reply]
It is nonsensical, because a kernel would not use *LIBC*, which is a userspace library.

**Reply by electrikjesus** [reply]
There is no kernel compiled with Waydroid source, as it is dependent on your host machine kernel and it's binder module

**Reply by roke-julian-lockhart** [reply]
> https://github.com/waydroid/waydroid/discussions/1552#discussioncomment-12544200

@aleasto, thanks for that. For others subscribed, [`stackoverflow.com/revisions/19490908/1`](https://stackoverflow.com/revisions/19490908/1#:~:text=There%20is%20actually%20no%20libc%20in%20kernel%20space.&text=But%20almost%20all%20functions%20from%20libc%20that%20make%20sense%20in%20kernel%20space%20are%20ported.&text=As%20far%20as%20I%20know%20these%20two%20implementations%20don't%20share%20codebase.) appears to accurately elaborate.
